src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala

```diff
@@ -0,0 +1,88 @@
+package beam.agentsim.infrastructure.parking
+
+import scala.language.higherKinds
+
+import cats.Monad
+
+import beam.agentsim.infrastructure.charging.ChargingPoint
+
+/**
+  * stores the number of stalls in use for a zone of parking stalls with a common set of attributes
+  *
+  * @param stallsAvailable a (mutable) count of stalls free (a semiphore)
```

Comment:
There's nothing I see that guarantees this as a semaphore, unless I'm missing something.

Reply:
In fact, I'm not 100% that the stall changing is even what would be considered pure...

Reply:
you're right, it's absolutely not a semaphore. i think i meant "it's designed like a semaphore in that it releases 'parking stalls' as it counts down from the max value, and no longer releases parking stalls when the stallsAvailable reaches 0." sort of as a note to communicate to the developer what it's doing logically.

purity, you mean FP? the m.pure wrapper here would allow us to run this in some context like cats.effect.IO if we wanted to in the future to run things in a more pure way. if we don't provide an effect, it will run as is without any context, so no one needs to worry about pure FP if they don't want to. cats was already in the dependencies so i went for it.

Reply:
also, since you mentioned concurrency: i've written this with the assumption that only one Actor has access to this data at a time, therefore the assumption that it doesn't need to consider synchronization issues.

use case: a ZonalParkingManager or RideHailManager will be the sole broker for parking stalls, and it will refer to an Array[ParkingZone] which tells it (on aggregate) the availability of parking resources (aka ParkingStalls).

Reply:
> purity, you mean FP? the m.pure wrapper here would allow us to run this in some context like cats.effect.IO if we wanted to in the future to run things in a more pure way. if we don't provide an effect, it will run as is without any context, so no one needs to worry about pure FP if they don't want to. cats was already in the dependencies so i went for it.

Admittedly my pure FP/cats is not strong. From reading about pure it seemed that it was meant to wrap pure code - and this clearly side effects.

> also, since you mentioned concurrency: i've written this with the assumption that only one Actor has access to this data at a time, therefore the assumption that it doesn't need to consider synchronization issues.

I kind of figured this might stay within 1 actor, but I guess I was figuring what makes sure nobody abuses that? When the mutation is encapsulated inside the actor then it's explicitly wrapped. Whereas this can escape should somebody not pay attention. I don't have a better solution to such a problem, though - and will remove the blocker. If interested I can ping the Akka team to see if there IS a solution for what I'm talking about?

Reply:
> Admittedly my pure FP/cats is not strong. From reading about pure it seemed that it was meant to wrap pure code - and this clearly side effects.

Sure, that's a confusing semantic. 

The purpose of any Monad.pure{} block is to wrap/lift the enclosed expression into an execution context or container. In pure FP, we want to control when side effects happen, and we do this by turning expressions with side-effects (impure) into something that will produce the same result every time you evaluate it (pure). So, this mutation would be run in some effect monad such as [cats.effect.IO](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/IO.scala). 

The "how" is that the expression would be placed in essentially a memoized thunk, deferring its execution until later and preventing multiple executions. This thunk would be evaluated when ZonalParkingManager wants to evaluate it. So, one way then to prevent race conditions on a mutation within a computational context (Actor in this case) is that the actor has a schedule to finally evaluate these mutations (stored in a concurrent collection type), and can perform them at a known phase of a procedure where it knows no reads/writes are happening. A nice feature of this abstraction is that it would also makes it easier to see where updates are happening, as well as running optimizations on the queue of operations stored.

I don't think anyone on this team is into the pure FP thing so I intentionally didn't go all in with this. The ParkingZone operations which mutate state are wrapped in a monad, but the code calling it isn't running it in a context, so it ends up running like a OO increment() op.

> I kind of figured this might stay within 1 actor, but I guess I was figuring what makes sure nobody abuses that?

I think the only way to enforce this via code is for the operations to be Actor-scoped, either in their parameters or in their visibility. But I think that it's clearer to have these things separated. My feeling is that we can't enforce it, but we can provide comments, and make sure that these attributes are private attributes of an Actor. The latter is currently the case as the `stalls` argument is not exposed (not prefixed with `val` in the class constructor):

```Scala
class ZonalParkingManager(
  val beamServices: BeamServices,
  stalls: Array[ParkingZone],
  searchTree: ParkingZoneSearch.StallSearch,
  rand: Random
) extends Actor
    with HasServices
    with ActorLogging { ... }
```